core/kazoo_number_manager/src/knm_port_request.erl

```diff
@@ -563,6 +594,66 @@ transition_numbers(PortReq) ->
             end
     end.
 
+-spec update_used_by(list(), 'undefined' | kz_term:ne_binary()) -> 'ok'.
+update_used_by(_Numbers, 'undefined') -> 'ok';
+update_used_by(Numbers, App) ->
+    lager:debug("update number ~p with used_by ~p", [Numbers, App]),
+    _Ok = knm_numbers:assign_to_app(Numbers, App),
+    'ok'.
+
+%%------------------------------------------------------------------------------
+%% @doc Apps used by in the port in document.
+%% This may not be a reliable source of used_by comparing to the callflow doc.
+%% @end
+%%------------------------------------------------------------------------------
+-spec app_used_by_portin(kz_term:ne_binary(), kz_json:object()) -> 'undefined' | list().
+app_used_by_portin(Numbers, JObj) ->
+    NumbersObj = kzd_port_requests:numbers(JObj),
+    [{Num, kz_json:get_value([Num, ?USED_BY_KEY], NumbersObj)} || Num <- Numbers].
+
+maybe_reconcile_app_used_by(Numbers, AccountId, JObj) ->
+    AccountDb = kz_util:format_account_db(AccountId),
+    PortInUsedBy = app_used_by_portin(Numbers, JObj),
+    AppUsedBy = used_by_which_app(AccountDb, Numbers),
+    _Ok=[maybe_reconcile_app(proplists:get_value(N, PortInUsedBy), App, N, JObj) || {N, App} <- AppUsedBy],
+    app_used_by(AppUsedBy).
+
+maybe_reconcile_app(App, App, _Num, _JObj) ->
+    {'ok', App};
+maybe_reconcile_app(_PortInApp, App, Num, JObj) ->
+    lager:warning("port in number ~p app ~p replaced by correct app ~p", [Num, _PortInApp, App]),
+    _Ok = assign_to_app(Num, App, JObj),
+    {'ok', App}.
+
+app_used_by([]) -> 'undefined';
+app_used_by([{_Num, App}|_T]) -> App.
```

Comment:
Any Number can be assigned to any app.

Group numbers by app in previous function. return it as result. After port doc is saved loop over proplist and assign numbers to their app.